File: kr3/kr3.2/kr3.2/priorityQueue.cpp
#include "priorityQueue.h"
#include <string>

using namespace std;

struct Node
{
	int priority = 0;
	std::string value = 0;
	Node *next{};
};

struct PriorityQueue
{
	Node *head{};
	Node *tail{};
};

PriorityQueue *createQueue()
{
	return new PriorityQueue{ nullptr };
}

bool isEmpty(PriorityQueue *queue)
{
	return !queue->head;
}

void enqueue(PriorityQueue *queue, int priority, string value)
{
	if (isEmpty(queue))
	{
		queue->head = new Node{ priority, value };
		queue->tail = queue->head;
		return;
	}

	queue->tail->next = new Node{ priority, value };
	queue->tail = queue->tail->next;
}

Node *searchMax(PriorityQueue *queue)
{
	auto node = queue->head;
	auto maxPriority = queue->head;
	while (node)
	{
		if (node->priority > maxPriority->priority)
		{
			maxPriority = node;
		}
		node = node->next;
	}
	return maxPriority;
}

string dequeue(PriorityQueue *queue)
{
	if (isEmpty(queue))
	{
		return "-1";
	}

	auto maxPriority = searchMax(queue);

	if (maxPriority == queue->head)
	{
		queue->head = maxPriority->next;
		string temp = maxPriority->value;
		delete maxPriority;
		return temp;
	}

	auto node = queue->head;
	auto prev = node;

	while (node)
	{
		if (node == maxPriority)
		{
			break;
		}
		prev = node;
		node = node->next;
	}

	if (maxPriority != queue->tail)
	{
		prev->next = node->next;
	}
	else
	{
		prev->next = nullptr;
		queue->tail = prev;
	}

	string temp = node->value;
	delete node;
	return temp;
}

void deleteQueue(PriorityQueue *queue)
{
	if (queue->head)
	{
		Node *current = queue->head;
		while (current)
		{
			Node *temp = current;
			current = current->next;
			delete temp;
		}
	}
	delete queue;
}

File: sem2/kr1/kr1/kr1/PriorityQueue.cs
namespace kr1
{
    /// <summary>
    /// Priority queue, an abstract container of (priority, value) pairs.
    /// </summary>
    public class PriorityQueue
    {
        /// <summary>
        /// An element of the priority queue.
        /// </summary>
        private class Node
        {
            public string Data { get; private set; }
            public Node Next { get; set; }
            public int Key { get; private set; }

            public Node(int newKey, string newData, Node newNext)
            {
                Data = newData;
                Key = newKey;
                Next = newNext;
            }
        }

        /// <summary>
        /// The element first added to the priority queue.
        /// </summary>
        private Node head = null;

        /// <summary>
        /// The element last added to the priority queue.
        /// </summary>
        private Node tail = null;

        /// <summary>
        /// Checks if the priority queue is empty.
        /// </summary>
        /// <returns>True if the priority queue is empty and false if it's not.</returns>
        public bool IsEmpty() => head == null;

        private void EnqueueWhenQueueIsEmpty(int key, string data)
        {
            head = new Node(key, data, head);
            tail = head;
        }

        /// <summary>
        /// Adds (priority, value) pair to the priority queue.
        /// </summary>
        /// <param name="key">An integer number, priority of the value.</param>
        /// <param name="data">A string data to add.</param>
        public void Enqueue(int key, string data)
        {
            if (IsEmpty())
            {
                EnqueueWhenQueueIsEmpty(key, data);
                return;
            }

            tail.Next = new Node(key, data, null);
            tail = tail.Next;
        }

        private Node SearchMax()
        {
            var node = head;
            var maxPriority = head;
            while (node != null)
            {
                if (node.Key > maxPriority.Key)
                {
                    maxPriority = node;
                }
                node = node.Next;
            }
            return maxPriority;
        }

        private string DequeueFromHead(Node maxPriority)
        {
            head = maxPriority.Next;
            return maxPriority.Data;
        }

        private Node FindPreviousNode(Node node)
        {
            var temp = head.Next;
            var previous = head;

            while (temp != null)
            {
                if (temp == node)
                {
                    return previous;
                }

                previous = temp;
                temp = temp.Next;
            }

            return null;
        }

        /// <summary>
        /// Removes value with maximum priority from the priority queue.
        /// </summary>
        /// <returns>A value with maximum priority.</returns>
        public string Dequeue()
        {
            if (IsEmpty())
            {
                throw new DequeueWhenQueueIsEmptyException();
            }

            var maxPriority = SearchMax();

            if (maxPriority == head)
            {
                return DequeueFromHead(maxPriority);
            }

            var previous = FindPreviousNode(maxPriority);

            if (maxPriority != tail)
            {
                previous.Next = maxPriority.Next;
            }
            else
            {
                previous.Next = null;
                tail = previous;
            }

            return maxPriority.Data;
        }
    }
}
